akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowPrefixAndDownstreamSpec.scala

```diff
@@ -0,0 +1,219 @@
+      val (prefixF, suffixF) = Source(0 until 100)
+        .prefixAndDownstreamMat(10) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix)
+        }(Keep.right)
+        .take(10)
+        .toMat(Sink.seq)(Keep.both)
+        .run
+
+      prefixF.futureValue should ===(0 until 10)
+      suffixF.futureValue /*(Interval(Span(1000, Minutes)))*/ should ===(10 until 20)
+    }
+
+    "propagate materialization failure" in assertAllStagesStopped {
+      val suffixF = Source(0 until 100)
+        .prefixAndDownstreamMat(10) { prefix =>
+          sys.error(s"I hate mondays! (${prefix.size})")
+        }(Keep.right)
+        .to(Sink.ignore)
+        .run
+
+      val ex = suffixF.failed.futureValue
+      ex.getCause should not be null
+      (ex.getCause should have).message("I hate mondays! (10)")
+    }
+
+    "propagate flow failures" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source(0 until 100)
+        .prefixAndDownstreamMat(10) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).map {
+            case 15 => sys.error("don't like 15 either!")
+            case n  => n
+          }
+        }(Keep.right)
+        .toMat(Sink.ignore)(Keep.both)
+        .run
+      prefixF.futureValue should ===(0 until 10)
+      val ex = suffixF.failed.futureValue
+      //ex.printStackTrace()
+      (ex should have).message("don't like 15 either!")
+    }
+
+    "produce multiple elements per input" in assertAllStagesStopped {
+      val (prefixF, suffixF) = src10()
+        .prefixAndDownstreamMat(7) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).mapConcat(n => List.fill(n - 6)(n))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should ===(0 until 7)
+      suffixF.futureValue should ===(7 :: 8 :: 8 :: 9 :: 9 :: 9 :: Nil)
+    }
+
+    "succeed when upstream produces no elements" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source
+        .empty[Int]
+        .prefixAndDownstreamMat(7) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).mapConcat(n => List.fill(n - 6)(n))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should be(empty)
+      suffixF.futureValue should be(empty)
+    }
+
+    "apply materialized flow's semantics when upstream produces no elements" in assertAllStagesStopped {
+      val (prefixF, suffixF) = Source
+        .empty[Int]
+        .prefixAndDownstreamMat(7) { prefix =>
+          Flow[Int].mapMaterializedValue(_ => prefix).mapConcat(n => List.fill(n - 6)(n)).prepend(Source(100 to 101))
+        }(Keep.right)
+        .toMat(Sink.seq[Int])(Keep.both)
+        .run()
+
+      prefixF.futureValue should be(empty)
+      suffixF.futureValue should ===(100 :: 101 :: Nil)
+    }
+
+    "handles upstream cancellation" in assertAllStagesStopped {
```

Comment:
Upstream does not cancel, it completes. Not entirely sure this test adds anything on top of the `Source.empty` one above.

Reply:
mostly right, but this example shows that the flow may produce even when upstream produces no elements at all.